### PR TITLE
Allow remote state reference in team_access team name

### PR DESCRIPTION
**Pull request by ryanwholey**
The problem:
Remote state references for team access is tough because the team name is the logical choice for the terraform resource name and it's also how we want to look up team access for importing. We can't easily look up the team ID if the team name is a reference to remote state. This differs from variable remote state reference where the value is an inconsequential property with regards to terraform indexing. 

For the time being, I think we should explicitly not support _importing_ remote state referenced teams. As a workaround, the team name can be used directly for import, then the reference can be used without issue from that point on. The proposed workflow does not require a merge to import, so this can be done with a single pr. 

Changes:
- Instead of teamName for the team_access resource name, use `team_acces_$n` with the ta index. This allows us to use remote references for the actual team name and team ID lookups while having a Terraform key that is unique and valid.
- Moves Import over to the TeamAccess struct instead of as a standalone function. I think these import functions make more sense on the types than in a generic import file, if you agree I'll move the others in a subsequent pr.

Other considerations:
- Using `count` to index the resources for the terraform plan instead of a unique resource with an index in the string name could work for solving the naming problem, but doesn't address the import issue of finding the team name. It also means a somewhat complex count dataset being stringified. It's complicated further by the option to pass either a generic access field or a set of permissions, so the values are a bit more complex (might not be exactly it, `access = ${count[i].value.access || null}`) 
- Using the API client to look up the team name and ID is not really feasible because its currently input as a remote state reference expression, which doesn't give us a name or an ID to use.
- A function that creates a small terraform workspace to fetch remote state, which we then use in the terraform plans. This is a bit tricky, as it would require some static analysis of the inputs to determine whether they are using remote state or not.

### CODE REVIEW

**Review by bendrucker — COMMENTED** [review]
Having an index in the resource name (or `count` for that matter) means that adding a new entry to the start of an existing list triggers a bunch of replacements. And maybe that's even racy if any of the creates happen before the resource with the old index is destroyed.

Seems like the optimal thing here would be to feed these in as a `map` to `for_each` with the team names as keys. 

> For the time being, I think we should explicitly not support importing remote state referenced teams

Probably ok. 

Alternatively, you could do 2 `plan` passes. The first would be analyzed to determine imports. This would reverse the amount of reverse engineering necessary to do imports, since the logic would become, "for all new resources in plan, check for existence, generate import string using attributes, attempt import." You'd create some sort of object where you could register resource types (e.g. `tfe_workspace`), along with config/functions for importing them. Like a function for checking whether it exists (should import) and one for generating the import ID from the planned state. 

This starts to fray a bit when resources are dependent on a computed property, e.g., things will depend on the workspace ID, which will be initially unknown. So perhaps it's actually 3 passes, first to import the WS, then to import all the dependent resources, last to generate a plan to be applied. Even if you leave `import: true` permanently, normally the initial plan will have no changes, and you can exit then. Only in the case of new workspaces are you planning 3x.

**Comment by ryanwholey** [comment]
Yeah good thought regarding the index, that would be unexpected behavior from the users perspective. How about a similar approach where we strip the team name input of interesting characters? 

The multi plan approach is interesting.. i'm wondering if there's any real boundary stopping the plans at 3? Seems unlikely, but in a contrived example and there were n dependents, we would have to run n + 1 (import) + 1 plan tf runs?

**Comment by bendrucker** [comment]
Re `n=3` you are correct that an arbitrary TF module could have a large but definite N. For this you could just hardcode those three passes based on the expected relationships/resources. Generalizing to a graph walking importer would be cool but challenging.

**Comment by ryanwholey** [comment]
stripping special characters from the input to use as a resource name [does plan without error](https://github.com/TakeScoop/terraform-provider-configuration/runs/3303232384?check_suite_focus=true), but it will be somewhat of a mess when it comes to importing. I had mentioned using the team name for actually importing and then moving back to the remote state lookup, but this wont work as smoothly because the resource name will change in this situation. In the indexed name situation, the resource name would have remained the same but here it is tied directly to the input so you'd be forced to do an additional state mv which I don't think is a good enough option for us.

**Comment by ryanwholey** [comment]
I'll go with a team access map input approach as you suggested and see how we like it. I'm trying to avoid going to multiple tf plans because I think it adds a decent amount of complexity but thats definitely still on the table

File: team_access.go [hunk 1]
```diff
@@ -30,3 +40,66 @@ func MergeWorkspaceIDs(teamAccess []TeamAccess, workspaces []*Workspace) []TeamA
...
+func (ta TeamAccess) GetResourceName() string {
+	//  Remove all non word characters
+	rx := regexp.MustCompile(`\W`)
+	return rx.ReplaceAllString(ta.TeamName, "")
+}
+
+// Import imports a team access resource by looking up an existing relation
+func (ta *TeamAccess) Import(ctx context.Context, tf *tfexec.Terraform, client *tfe.Client, organization string, opts ...tfexec.ImportOption) error {
+	if strings.HasPrefix("${", ta.TeamName) {
+		return fmt.Errorf("importing remote state team access references is not currently supported, use real team name and revert if desired")
```
Comment (bendrucker) [inline team_access.go:54]:
```suggestion
		return fmt.Errorf("Team access only supports importing statically named teams, but an expression (${}) was found. Convert the team name to a static string to allow import.")
```

File: team_access.go [hunk 2]
```diff
@@ -30,3 +40,66 @@ func MergeWorkspaceIDs(teamAccess []TeamAccess, workspaces []*Workspace) []TeamA
 
 	return ts
 }
+
+// GetResourceName returns the resource name used for the team and access resources in the Terraform configuration
+func (ta TeamAccess) GetResourceName() string {
```
Comment (bendrucker) [inline team_access.go:45]: This is solved by `for_each` which has no restrictions on whitespace in keys

Comment (bendrucker) [reply team_access.go:45]: There are other invalid chars for resource names as well, e.g. `.`

**Comment by bendrucker** [comment]
The `map` is important on the TF side but you don't need to change the action interface. You can convert the list of objects in Go to a `map[string][string]` and feed that in as a var.

**Comment by ryanwholey** [comment]
While I work on the for_each version of this, this version _does_ require the user to define the teams as a map in the Github action interface. It's pretty straight forward and should work for importing static names.

One of the things I'm working through with the for_each approach is an elegant way to use either access or permissions for the team access block in the for each. We will see how it goes.

**Comment by bendrucker** [comment]
Use dynamic blocks for `permissions` support to conditionally create 0 or 1 blocks

**Comment by ryanwholey** [comment]
okay! here's an implementation using a for each approach. 

This approach still has the same problem that removing illegal keys from the value inputs has, and that is that when you change the team_name reference to be the real team name and import, the resource address will be different when you change the team_name back to the reference.

I think the most straight forward approach here would be to have the user pass the the team names as a map, (the previous commit) because it allows for importing at the same address. 

Plus all the for_each stuff is a bit hard to follow. I think the inconvenience to the user of adding a unique key for each team access resources outweighs the complexity of this.

I also think I'll move the import back to import.go for now as it just adds noise to this pr

**Comment by bendrucker** [comment]
Asking users to provide stable map keys leaks out TF implementation details

Took me a re-read to figure this out and what I was missing:

> This approach still has the same problem that removing illegal keys from the value inputs has, and that is that when you change the team_name reference to be the real team name and import, the resource address will be different when you change the team_name back to the reference.

This is about changing a _name_ to an _ID_. The fix here is to always resolve the team name to ID and use those as `for_each` keys. Allowing names _and_ IDs optionally via expressions makes this hard. IDs should be easily detectable unless they're expressions. Splitting to `team_name` and `team_id` would make that simple.

**Comment by ryanwholey** [comment]
Might be missing your point here, but here's how I see it:

If we allow expressions in team names, we'll need to do some preprocessing to convert the name to an ID. This will end up being a separate terraform apply/output that we would map back to the expression.

Agreed that with a team ID, we can use that in the for_each or we can ditch the for_each all together because IDs should be valid as resource names. 

Accepting IDs _and_ names doesn't really help solve this problem for the case where people prefer to use names. 

Accepting only IDs removes the need for the team name altogether, but isn't so nice for folks who use literal team IDs directly instead of a reference with human friendly map key names

Here are the possible approaches as I see it:
- Only support team_id, showing an example using remote state in the docs
- Not support team_name expressions
- Run an apply on remote state blocks to fetch all output and use that to construct the workspace plan

Might be missing what you were saying with splitting team name and team id?

**Comment by ryanwholey** [comment]
thinking further, if we do teamIDs only and no preprocessing, we would still want the for_each to account for the "expression as a resource name" problem. but passing team IDs certainly simplifies the logic a fair amount

**Comment by bendrucker** [comment]
> If we allow expressions in team names, we'll need to do some preprocessing to convert the name to an ID. 

`data.tfe_team` will handle in the same run. My thought was that you need a way to distinguish whether the user is trying to set a name or an ID to determine which name -> ID translations you need to do. But on further reflection even that is not necessary. Terraform will resolve the expressions whether they're names or IDs and then from there you can decide what kind of input was passed. If you set `local.teams` as a map, where the keys are the `team` values from the list and the values are the access/permissions:

```tf
data "tfe_team" "teams" {
  for_each = [ for k keys(local.teams) : k if length(regexall("^team-", k)) > 0 ]

  name = each.key
  organization = "..."
}
```

Then, in the `tfe_team_access` resource, perform the same regex logic checking if `each.key` is a team ID. If not, reference `data.tfe_team.teams[each.key].id`.

> Only support team_id, showing an example using remote state in the docs

This only makes sense if you use `tfe_team` to create teams somewhere else. Users may want to manage their teams some other way.

**Comment by bendrucker** [comment]
Err, not using `each.key` actually, this has to happen before that to produce the right address. You'd do it in the `for_each` expression in `tfe_team_access` itself. You'd have another `for` comprehension there iterating through a `map(string)` and in some cases resolving to a different key, if the the key is not team-ID-like

**Comment by ryanwholey** [comment]
Alright! This iteration uses incorporates the team ID into the resource ID.

If `team_name` is passed, team ID is referenced through a data `tfe_team` block that looks up that team ID.
If `team_id` is passed, the team ID is used directly, and no `tfe_team` data block is created for that team.
`team_id` takes precedence over `team_name`.

When importing, a static team_id _must_ be passed. Once imported, the user can use expressions for team ID/team names, or literal team IDs/team names.

I think this covers all of our requirements and with a reasonable need for a static `team_id` when importing.

File: README.md [hunk 3]
```diff
@@ -128,23 +128,49 @@ with:
...
+        region: us-east-1
+```
+
+To import existing team access resources, a static value for `team_id` must be supplied
+
+```yml
+with:
+  import: true
+  team_access: |-
+    - team_name: Readers
```
Comment (bendrucker) [inline README.md:162]: Why allow both `team_name` and `team_id`? I suggested splitting with the thought that it was necessary for disambiguating IDs from names, but TF will resolve any expressions and `^team-` will identify team IDs.

If a name _and_ an ID are supplied but they don't agree, the correct behavior would be indeterminate. The only safe thing to do is to error and exit.

You could support both for clarity, but I think just `team` that is either an ID or a name should work fine.

**Review by bendrucker — COMMENTED** [review]
Sweet! One remaining question about accepting both name + ID, then this looks good to go. I can take a stab at the multi-stage plan approach for computing importable resources tomorrow/Monday.

File: README.md (hunk 3 above)
Comment (ryanwholey) [reply README.md:162]: Yeah i can poke at that. My thought was that making the documentation clear that ID takes precedence over name would be a reasonable. 

In practice it felt like a nice experience to just throw the ID in there for import and then delete it without having to change and change back

Comment (bendrucker) [reply README.md:162]: Avoid leaning too heavily on docs or coupling tools too tightly to your own workflow instincts. It is more important that software be correct (or predictable) than convenient, in the long run.

> In practice it felt like a nice experience to just throw the ID in there for import and then delete it without having to change and change back

A code comment handles this without creating the opportunity for ambiguous/invalid inputs.

Comment (bendrucker) [reply README.md:162]: The prospect of importing by static name and switching to an ID expression in a future commit is awkward. Too cumbersome across 10s or 100s of repos. It's good that it's _possible_ but don't worry about making it _optimal_. Save that energy for making imports work without caveats.

Comment (ryanwholey) [reply README.md:162]: ah i knew there was something but I couldn't quite remember

If both name and ID input under the same key, I don't believe there's a way to tell whether an expression points to a team ID or a team name. This is important to determine whether to add a the team data block and use that as a reference or not.

How about we error if they're both supplied?

Comment (bendrucker) [reply README.md:162]: > If both name and ID input under the same key, I don't believe there's a way to tell whether an expression points to a team ID or a team name. This is important to determine whether to add a the team data block and use that as a reference or not.

This was my original thought. But that's only true if you're thinking of the inputs as static strings, e.g., in Go. At the stage of runtime we're talking about, Terraform only cares if values are _known_. Any expression that is known can be evaluated and used in `for_each`. 

Team IDs have a prescribed format (`team-*`), that can be used to determine which inputs are team IDs versus names:

https://github.com/TakeScoop/terraform-cloud-workspace-action/pull/30#issuecomment-897252999

Anything that is not a team ID could be assumed to be a team name. I just checked though, and that assumption isn't strictly true. If you want, you can create a team named `team-foo`, or even copy the exact ID of another team as a name. So...

> How about we error if they're both supplied?

This seems like the best option. Turns out you really do need additional info to unambiguously identify the team after all.

Minor:

```yaml
team: # either or
  id: team-abc123
  name: Foo
```

Nesting versus prefixing guarantees that `name` and `id` will be next to each other so the source of the conflict is clear. Not particularly important given that team access objects have 4 possible top level keys. Would have a more appreciable impact if you had 20 keys and `team_id` and `team_name` could end up physically far from each other.

Comment (ryanwholey) [reply README.md:162]: I see, yes we could have pushed that is-team-id check within terraform instead of in Go, and let Terraform conditionally construct the tfe_team block and the team access team ID ref. 

Also good point regarding team names as IDs.

Comment (bendrucker) [reply README.md:162]: Yep, anything where you're going to allow an expression, you cannot consume the value in Go, only pass it.

**Review by bendrucker — APPROVED** [review]
Nicely done!
